code/utils.py

```diff
@@ -131,3 +135,120 @@ def viz_class_video(fpath: pathlib.Path):
         video.write(cv2.resize(cv2.imread(str(image)), (width, height)))
     cv2.destroyAllWindows()
     video.release()
+
+
+##################
+# h3 funcs
+##################
+
+
+def make_grid(fua, res, proj_crs):
+    """
+    create a hex grid for given FUA (using original OSM dataset) and resolution.
+    fua: int, ID of city (any of use cases: 809, 869, 1133, 1656, 4617, 4881, 8989)
+    res: int, hex grid resolution
+    (suggested: 8 - with cell side length ~530m or 9 - cell side length ~200m)
+    """
+
+    # CRS for h3 must be 4326
+    orig_crs = "EPSG:4326"
+
+    # get city geom and name
+    meta = geopandas.read_parquet("../data/sample.parquet")
+    geom = meta.loc[meta.eFUA_ID == fua, "geometry"]
+
+    # read in OSM data
+    orig = geopandas.read_parquet(f"../data/{fua}/roads_osm.parquet")
+    orig = orig.to_crs(orig_crs)
+
+    # create h3 grid
+    temp = mapping(geom)["features"][0]["geometry"]
+    temp["coordinates"] = [
+        [[j[1], j[0]] for j in i] for i in temp["coordinates"]
+    ]  # reverse coords
+    hex_pd = pd.DataFrame(list(h3.polyfill(temp, res=res)), columns=["hex_id"])
+    hex_pd["geometry"] = [
+        Polygon(h3.h3_to_geo_boundary(x, geo_json=True)) for x in hex_pd["hex_id"]
+    ]
+    grid = geopandas.GeoDataFrame(hex_pd)
+    grid.set_crs(orig_crs, inplace=True)
+
+    # keep only the grid cells that contain some piece of the orig data
+    mytree = strtree.STRtree(geoms=orig.geometry)
+    q = mytree.query(grid.geometry, predicate="intersects")
+    grid = grid.loc[sorted(set(q[0]))]
```

Comment:
geopandas `.sindex` has a `sort` keyword in the `query` method, so you don't need to do that yourself.

Reply:
... and likewise